test/conformance/header_test.go

```diff
@@ -0,0 +1,104 @@
+// +build e2e
+
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conformance
+
+import (
+	"net"
+	"regexp"
+	"strings"
+	"testing"
+
+	"github.com/knative/pkg/test/logging"
+	"github.com/knative/serving/test"
+)
+
+// TestMustHaveHeadersSet verified that all headers declared as "MUST" in the runtime
+// contract are present from the point of view of the user container.
+func TestMustHaveHeadersSet(t *testing.T) {
+	logger := logging.GetContextLogger(t.Name())
+	clients := setup(t)
+
+	ri, err := fetchRuntimeInfo(clients, logger, &test.Options{})
+	if err != nil {
+		t.Fatalf("Error fetching runtime info: %v", err)
+	}
+
+	// For incoming requests, the Host header is promoted to the
+	// Request.Host field and removed from the Header map. Therefore we
+	// check against the Host field instead of the map.
+	if ri.Request.Host == "" {
+		// We just check that the host header exists and is non-empty as the request
+		// may be made internally or externally which will result in a different host.
+		t.Error("Header host was not present on request")
+	}
+}
+
+// TestMustHaveHeadersSet verified that all headers declared as "SHOULD" in the runtime
+// contract are present from the point of view of the user container.
+func TestShouldHaveHeadersSet(t *testing.T) {
+	logger := logging.GetContextLogger(t.Name())
+	clients := setup(t)
+
+	expectedHeaders := map[string]*regexp.Regexp{
+		// We expect the protocol to be http for our test image.
+		"x-forwarded-proto": regexp.MustCompile("https?"),
+		// We expect the value to be a list of at least one comma separated IP addresses (IPv4 or IPv6).
+		// We do additional validation of the IP addresses below.
+		"x-forwarded-for": regexp.MustCompile("[0-9a-f.:]*(, [0-9a-f.:]*)*"),
+
+		// Trace Headers
```

Comment:
Are these set in the request? The runtime spec is attempting to encourage forwarding all headers unless there are specific reasons not to:

> all inbound request headers SHOULD be set to the same values as the incoming request.\

I can see where the tracing section is ambiguous; we should clarify that tracing headers may not be set on all requests.

Reply:
FYI there is `TestMust` and `TestShould`, this is the latter.

Reply:
I may be misunderstanding your question. 

Tracing headers are not set on the initial client request but are injected in via Zipkin in Istio.

None of these headers are set by the client (which is a gap in the test coverage I can address).

Reply:
Created #3208 as setting headers is going to take updating knative/pkg. I will pick this up as a follow-up to this change.

Reply:
The runtime contract in this section is attempting to ensure that distributed tracing works when Knative is part of the overall service, i.e.

```
external system ---> knative
```

Additionally, this section somewhat clumsily references RFC7320; in particular, the `Host` and `Connection` headers have specific HTTP/1.1 behavior.